### DISCUSSION TITLE
VTSessionDB vs. Program object locking

### DISCUSSION THREAD

**Discussion by v-p-b** [NationalSecurityAgency/ghidra#7044] [Q&A]
I'm trying to use pyhidra to initiate a Version Tracking session, my code boils down to this:

```py

p = GhidraProject.openProject(sys.argv[1], sys.argv[2], False)
# ...    
src_program = p.openProgram("/".join(src_path[0]), src_path[1], False)
dst_program = p.openProgram("/".join(dst_path[0]), dst_path[1], False)
# ...
db = VTSessionDB("Test", src_program, dst_program, session_ref)
```

On `VTSessionDB` instantiation I get the following exception:

```
java.lang.RuntimeException: java.lang.RuntimeException: ghidra.framework.store.LockException: domain object(s) are busy/locked
```

`.isLocked()` on both `src` and `dst` Program objects return False. I don't have other running Ghidra processes either.

How should I create input objects for `VTSessionDB` so locking won't break the process?

**Comment by ryanmkurtz** [comment]
I'm playing around with this now, and am having the same issue.  I am assuming there is an issue in `GhidraProject` class, which may not get exercised much these days.  In trying to use alternate API calls, it is clear that PyGhidra really needs some built-in support for working with projects better.  I plan on pursuing this in the coming weeks.

**Comment by v-p-b** [comment]
Thank you for looking into this!

While accessing Programs from a GhidraProject looks like the "right" way to me (Programs are part of Projects) I'm also interested in accessing Program objects in alternative ways, even from Java directly.

**Reply by ryanmkurtz** [reply]
Yes, it indeed does look like the right way. I need to ask other team members about that error tho.

If you are writing a more standard Java-based GhidraScript, you can see some examples here:
* https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/VersionTracking/ghidra_scripts/AutoVersionTrackingScript.java
* https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/VersionTracking/ghidra_scripts/CreateAppliedExactMatchingSessionScript.java
